clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/PatchExecutorSpec.scala

```diff
@@ -0,0 +1,177 @@
+
+    documents.foreach { km =>
+      val (theKey, metadata) = km
+      (
+        webClient
+          .upsert(_: UpsertAux[ArraysKey, ArraysMetadata])(
+            _: ArraysKey,
+            _: ArraysMetadata,
+            _: Boolean
+          )
+        )
+        .expects(ArraysIndex, theKey, metadata, false)
+        .returning(Source.single(UpsertId.nextId().asJson))
+    }
+  }
+
+  it should "not patch documents that already have metadata defined" in {
+    val webClient = mock[ClioWebClient]
+
+    val documentsWithMetadata = zipped.take(alreadyHasMetadata).map { km =>
+      val (key, metadata) = km
+      (key, metadata.copy(sampleAlias = newMetadata.sampleAlias))
+    }
+    val documentsWithoutMetadata = zipped.drop(alreadyHasMetadata)
+
+    setupQueryReturn(webClient, documentsWithMetadata ++ documentsWithoutMetadata)
+    setupUpserts(webClient, documentsWithoutMetadata.map(km => (km._1, newMetadata)))
+
+    val executor = new PatchExecutor(PatchArrays(loc))
+    executor.execute(webClient, getMockIoUtil()).runWith(Sink.seq).map { _ =>
+      succeed
+    }
+  }
+
+  it should "patch documents that don't have metadata defined" in {
+    val webClient = mock[ClioWebClient]
+
+    setupQueryReturn(webClient, zipped)
+    setupUpserts(webClient, keys.map(k => (k, newMetadata)))
+
+    val executor = new PatchExecutor(PatchArrays(loc))
+    executor.execute(webClient, getMockIoUtil()).runWith(Sink.seq).map { _ =>
+      succeed
+    }
+  }
+
+  it should "patch documents that have partial metadata" in {
+    val twoMetadata = newMetadata.copy(chipType = Some("a_chip_type"))
+    val ioUtil = mock[IoUtil]
+    (ioUtil
+      .readMetadata(_: ArraysIndex.type)(_: URI))
+      .expects(*, loc)
+      .returning(Source.single(twoMetadata))
+
+    val webClient = mock[ClioWebClient]
+
+    val returnedByQuery = zipped.map { km =>
+      val (key, metadata) = km
+      (key, metadata.copy(chipType = Some("some_other_chip_type")))
+    }
+
+    setupQueryReturn(webClient, returnedByQuery)
+    setupUpserts(webClient, keys.map(k => (k, newMetadata)))
+
+    val executor = new PatchExecutor(PatchArrays(loc))
+    executor.execute(webClient, ioUtil).runWith(Sink.seq).map { _ =>
+      succeed
+    }
+  }
+
+  it should "not upsert anything if nothing needs to be updated" in {
+    val ioUtil = mock[IoUtil]
+    (ioUtil
+      .readMetadata(_: ArraysIndex.type)(_: URI))
+      .expects(*, loc)
+      .returning(Source.single(ArraysMetadata()))
+
+    val webClient = mock[ClioWebClient]
+
+    setupQueryReturn(webClient, zipped)
+
+    val executor = new PatchExecutor(PatchArrays(loc))
+    executor.execute(webClient, ioUtil).runWith(Sink.seq).map { _ =>
+      succeed
+    }
+  }
+
+  it should "throw an exception if the wrong type of metadata is given" in {
+    recoverToSucceededIf[ClassCastException] {
+      val ioUtil = mock[IoUtil]
+      (ioUtil
+        .readMetadata(_: GvcfIndex.type)(_: URI))
+        .expects(*, loc)
```

Comment:
This `*` is the problem. `PatchExecutor` will always pass in the index for its command (`ArraysIndex` in this case), so this is an impossible expectation.

It should instead expect an `ArraysIndex` but return a `Source.failed` simulating one of the [possible](https://github.com/broadinstitute/clio/pull/267/files#diff-c2407722b19c3ff72071ca32bed8428fR58) [failures](https://github.com/broadinstitute/clio/pull/267/files#diff-c2407722b19c3ff72071ca32bed8428fR66) in `IoUtil`.